contributors/design-proposals/code-generator/kubegen.md

```diff
@@ -0,0 +1,333 @@
+- Run `kubegen version` to display the version
+  - Will correspond to a Kubernetes release - e.g. 1.8, 1.9
+
+## Generation for Versioned vs Internal types
+
+Code generation comes in 2 modes:
+
+- Generating clients and libraries for versioned APIs used by controllers, clis, and other clients.
+- Generating infrastructure for Kubernetes apiserver machinery to implement and expose API types internally
+
+If you are using CRDs for type definitions, then you should ignore internal types, and use the
+*versioned* commands.
+
+If *are* building an API directly into a Kubernetes API server or
+Kubernetes extension apiserver, then
+you will need to use the *internal* commands.  You may also want to use the *versioned*
+commands to publish libraries for your types externally - e.g. `k8s.io/client-go`
+
+### Running the command line
+
+#### Generate clients and related libraries for versioned (external) types
+
+```sh
+kubegen versioned
+```
+
+- run code generators against discovered API type definitions
+- prepend LICENSE to generated files (wrapping in comments)
+
+#### Generate clients and related libraries for internal types
+
+**Note:** This will *also* generate libraries for versioned types as internal types require
+versioned types to be useful.
+
+```sh
+kubegen internal
+```
+
+- run code generators against discovered API type definitions
+- prepend LICENSE to generated files (wrapping in comments)
+
+### Options
+
+A number of options can be specified to override the default behavior
+
+#### Define copyright owners
+
+The copyright owners for the license can be defined with the `--copyright` flag.
+This will prepend the license header at the top of generated files with a copyright string.
+e.g. *Copyright 2018 The Kubernetes Authors.*
+
+
+```sh
+kubegen versioned --copyright "The Kubernetes Authors."
+```
+
+#### Define the license file
+
+The license file can be overridden from the LICENSE file using the `--license-file` flag.
+This will use the contents from the boilerplate.go.txt file instead of LICENSE.
+If boilerplate.go.txt is already in comments then use it verbatim, otherwise wrap it in comments.
+
+
+**Note**: This passes the `--go-header-file` to each of the code generators
+
+```sh
+kubegen versioned --license-file "boilerplate.go.txt"
+```
+
+#### Define the license type
+
+If no license file exists, one may be specified on the using the `--license` flag.
+
+```sh
+kubegen versioned --license "Apache 2.0"
+```
+
+Options:
+
+- "Apache 2.0"
+- "None"
+
+#### Define API groups and versions
+
+The set of APIs code generation is performed for can be explicitly defined instead of
+implicitly determined using the `--api-version` and `--api-group` flags.  When these
+flags are provided, API groups will not be implicitly discovered.  Only API versions
+for groups specified with `--api-group` will be discovered.
+
+```sh
+kubegen versioned --api-version apps/v1 --api-version apps/v1beta1 --api-version extensions/v1beta1
```

Comment:
Why not `kubegen versioned apps/v1 apps/v1beta1 extensions/v1beta1` ?

Reply:
I thought about that.  I don't really like how it interacts with subcommands.  e.g. if we want to introduce specific subcommands for versioned, this will get really ugly.